File: lab2/B.cpp
#include <iostream>
#include <vector>
#include <cmath>      // For sqrt, abs
#include <iomanip>    // For std::fixed, std::setprecision
#include <algorithm>  // For std::min, std::max
#include <utility>    // For std::pair

struct Point3D {
    double x = 0.0, y = 0.0, z = 0.0;

    // Operator overloading for vector math
    Point3D operator+(const Point3D& other) const {
        return {x + other.x, y + other.y, z + other.z};
    }

    Point3D operator-(const Point3D& other) const {
        return {x - other.x, y - other.y, z - other.z};
    }

    Point3D operator*(double scalar) const {
        return {x * scalar, y * scalar, z * scalar};
    }

    Point3D operator/(double scalar) const {
        return {x / scalar, y / scalar, z / scalar};
    }

    double dot(const Point3D& other) const {
        return x * other.x + y * other.y + z * other.z;
    }

    double norm() const {
        return std::sqrt(x * x + y * y + z * z);
    }
};

// Segment struct
struct Segment {
    Point3D start;
    Point3D end;
};

double distance_point_to_segment(const Point3D& point, const Segment& segment) {
    Point3D p = point;
    Point3D a = segment.start;
    Point3D b = segment.end;

    Point3D ab = b - a;
    Point3D ap = p - a;

    double ab_len_sq = ab.dot(ab);

    // If segment is just a point
    if (ab_len_sq == 0) {
        return ap.norm();
    }

    // Project p onto the line ab, but clamp t to [0, 1]
    double t = ap.dot(ab) / ab_len_sq;
    t = std::max(0.0, std::min(1.0, t));

    Point3D closest = a + ab * t;
    return (p - closest).norm();
}

double compute_total_distance(const Point3D& point, const std::vector<Segment>& segments) {
    double total_dist = 0.0;
    for (const auto& seg : segments) {
        total_dist += distance_point_to_segment(point, seg);
    }
    return total_dist;
}

Point3D compute_gradient(const Point3D& point, const std::vector<Segment>& segments) {
    const double epsilon = 1e-6;
    double current_distance = compute_total_distance(point, segments);

    // Gradient for x
    Point3D point_dx = {point.x + epsilon, point.y, point.z};
    double dist_dx = compute_total_distance(point_dx, segments);
    double grad_x = (dist_dx - current_distance) / epsilon;

    // Gradient for y
    Point3D point_dy = {point.x, point.y + epsilon, point.z};
    double dist_dy = compute_total_distance(point_dy, segments);
    double grad_y = (dist_dy - current_distance) / epsilon;

    // Gradient for z
    Point3D point_dz = {point.x, point.y, point.z + epsilon};
    double dist_dz = compute_total_distance(point_dz, segments);
    double grad_z = (dist_dz - current_distance) / epsilon;

    return {grad_x, grad_y, grad_z};
}

std::pair<Point3D, double> gradient_descent(const std::vector<Segment>& segments) {
    // Hyperparameters
    const int max_iterations = 100000;
    const double learning_rate = 0.001;
    const double tolerance = 1e-6;

    // Initialize at centroid of all segment endpoints
    Point3D sum_points; // Initializes to (0,0,0)
    int n = 0;
    for (const auto& seg : segments) {
        sum_points = sum_points + seg.start;
        sum_points = sum_points + seg.end;
        n += 2;
    }

    Point3D point = sum_points / n;
    double prev_distance = compute_total_distance(point, segments);

    // Gradient descent loop
    for (int iteration = 0; iteration < max_iterations; ++iteration) {
        Point3D gradient = compute_gradient(point, segments);
        point = point - (gradient * learning_rate);

        double current_distance = compute_total_distance(point, segments);

        // Check for convergence
        double distance_change = std::abs(current_distance - prev_distance);
        if (distance_change < tolerance) {
            break;
        }

        prev_distance = current_distance;
    }

    double min_distance = compute_total_distance(point, segments);
    return {point, min_distance};
}

int main() {
    int n;
    std::cin >> n;

    std::vector<Segment> segments(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> segments[i].start.x >> segments[i].start.y >> segments[i].start.z
                 >> segments[i].end.x   >> segments[i].end.y   >> segments[i].end.z;
    }

    // Run gradient descent
    std::pair<Point3D, double> result = gradient_descent(segments);
    double min_distance = result.second;

    // Output result with 4 decimal places
    std::cout << std::fixed << std::setprecision(4) << min_distance << std::endl;

    return 0;
}

File: lab2/B.py
'''
Center of Segments
Limits: 2 sec., 512 MiB
Given a set of segments. You need to find a point such that the sum of distances from it to all segments is minimal.
Input
The first line contains one integer N - the number of segments.
The next N lines contain 6 integers separated by spaces - x, y, z coordinates of the start and x, y, z coordinates of the end of the segment.
Output
One number - the sum of distances from the found point to the given segments with precision up to 4 decimal places.
Constraints

1 ≤ N ≤ 100
-100 ≤ x, y, z ≤ 100
'''
from dataclasses import dataclass
import math


@dataclass
class Point3D:
    x: float
    y: float
    z: float

    def __add__(self, other):
        return Point3D(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other):
        return Point3D(self.x - other.x, self.y - other.y, self.z - other.z)

    def __mul__(self, scalar):
        return Point3D(self.x * scalar, self.y * scalar, self.z * scalar)

    def __truediv__(self, scalar):
        return Point3D(self.x / scalar, self.y / scalar, self.z / scalar)

    def dot(self, other):
        return self.x * other.x + self.y * other.y + self.z * other.z

    def norm(self):
        return math.sqrt(self.x ** 2 + self.y ** 2 + self.z ** 2)


@dataclass
class Segment:
    start: Point3D
    end: Point3D


def distance_point_to_segment(point, segment):
    """
    Calculate minimum distance from point to line segment in 3D
    """
    p = point
    a = segment.start
    b = segment.end

    ab = b - a
    ap = p - a

    ab_len_sq = ab.dot(ab)

    if ab_len_sq == 0:
        return ap.norm()

    t = ap.dot(ab) / ab_len_sq
    t = max(0, min(1, t))

    closest = a + ab * t

    return (p - closest).norm()


def compute_total_distance(point, segments):
    """
    Compute total distance from point to all segments
    """
    return sum(distance_point_to_segment(point, seg) for seg in segments)


def compute_gradient(point, segments, epsilon=1e-6):
    """
    Compute numerical gradient of the objective function
    """
    current_distance = compute_total_distance(point, segments)

    # Numerical gradient for x
    point_dx = Point3D(point.x + epsilon, point.y, point.z)
    dist_dx = compute_total_distance(point_dx, segments)
    grad_x = (dist_dx - current_distance) / epsilon

    # Numerical gradient for y
    point_dy = Point3D(point.x, point.y + epsilon, point.z)
    dist_dy = compute_total_distance(point_dy, segments)
    grad_y = (dist_dy - current_distance) / epsilon

    # Numerical gradient for z
    point_dz = Point3D(point.x, point.y, point.z + epsilon)
    dist_dz = compute_total_distance(point_dz, segments)
    grad_z = (dist_dz - current_distance) / epsilon

    return Point3D(grad_x, grad_y, grad_z)


def gradient_descent(segments, max_iterations=100000, learning_rate=0.001, tolerance=1e-6):
    """
    Find optimal point using gradient descent with distance-based convergence
    """
    # Initialize at centroid of all segment endpoints
    all_points = []
    for seg in segments:
        all_points.append(seg.start)
        all_points.append(seg.end)

    sum_x = sum(p.x for p in all_points)
    sum_y = sum(p.y for p in all_points)
    sum_z = sum(p.z for p in all_points)
    n = len(all_points)

    point = Point3D(sum_x / n, sum_y / n, sum_z / n)

    # Calculate initial distance
    prev_distance = compute_total_distance(point, segments)

    # Gradient descent loop
    for iteration in range(max_iterations):
        # Compute gradient
        gradient = compute_gradient(point, segments)

        # Update point
        point = point - gradient * learning_rate

        # Calculate new distance
        current_distance = compute_total_distance(point, segments)

        # Check for convergence based on distance change
        distance_change = abs(current_distance - prev_distance)
        if distance_change < tolerance:
            break

        # Update previous distance for next iteration
        prev_distance = current_distance

    # Calculate final minimum distance
    min_distance = compute_total_distance(point, segments)

    return point, min_distance


def main():
    n = int(input())
    segments = []

    for _ in range(n):
        coords = list(map(int, input().split()))
        start = Point3D(coords[0], coords[1], coords[2])
        end = Point3D(coords[3], coords[4], coords[5])
        segments.append(Segment(start, end))

    # Run gradient descent
    optimal_point, min_distance = gradient_descent(segments)

    # Output result with 4 decimal places
    print(f"{min_distance:.4f}")


if __name__ == "__main__":
    main()